File: Sources/Whispr/Dictionary.swift
import Foundation
#if canImport(AppKit)
import AppKit
#endif

struct Replacement: Codable, Identifiable, Equatable {
    var id = UUID()
    var from: String
    var to: String
}

struct DictionaryData: Codable, Equatable {
    var vocab: [String] = []              // preferred spellings, fuzzy-matched (e.g. "WhisperKit")
    var replacements: [Replacement] = []  // exact phrase -> replacement
}

/// Applies a user dictionary to a transcript: exact phrase replacements, then fuzzy correction
/// of near-miss words toward the user's custom vocabulary (Jaro-Winkler).
enum DictionaryStore {
    static let fuzzyThreshold = 0.86       // words of 6+ characters
    static let shortWordThreshold = 0.92   // 3-5 characters: one edit is a large share of a short word
    static let phraseThreshold = 0.92      // two spoken words merged into one vocab term

    static let url: URL = {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("OpenWispr", isDirectory: true)
        try? FileManager.default.createDirectory(at: base, withIntermediateDirectories: true)
        return base.appendingPathComponent("dictionary.json")
    }()

    static func load() -> DictionaryData {
        guard let data = try? Data(contentsOf: url),
              let d = try? JSONDecoder().decode(DictionaryData.self, from: data) else { return DictionaryData() }
        return d
    }

    static func save(_ d: DictionaryData) {
        if let data = try? JSONEncoder().encode(d) { try? data.write(to: url, options: .atomic) }
    }

    static func apply(_ text: String, _ d: DictionaryData,
                      isRealWord: (String) -> Bool = defaultIsRealWord) -> String {
        var s = text
        // 1) exact phrase replacements (case-insensitive)
        for r in d.replacements where !r.from.isEmpty {
            s = s.replacingOccurrences(
                of: "\\b\(NSRegularExpression.escapedPattern(for: r.from))\\b",
                with: r.to, options: [.regularExpression, .caseInsensitive]
            )
        }
        // 2) fuzzy-correct words toward custom vocab
        guard !d.vocab.isEmpty else { return s }
        let words = s.split(separator: " ", omittingEmptySubsequences: false).map(String.init)
        var out: [String] = []
        var i = 0
        while i < words.count {
            // Prefer a two-word match, so "open whisper" beats correcting each half separately.
            if i + 1 < words.count,
               let merged = mergedMatch(words[i], words[i + 1], vocab: d.vocab, isRealWord: isRealWord) {
                out.append(merged)
                i += 2
                continue
            }
            out.append(correctWord(words[i], vocab: d.vocab, isRealWord: isRealWord))
            i += 1
        }
        return out.joined(separator: " ")
    }

    private static func correctWord(_ word: String, vocab: [String], isRealWord: (String) -> Bool) -> String {
        // preserve leading/trailing punctuation
        let trimSet = CharacterSet(charactersIn: ".,!?;:\"'()")
        let core = word.trimmingCharacters(in: trimSet)
        guard core.count >= 3 else { return word }
        let lower = core.lowercased()
        // already correct (case-insensitive exact) → snap to vocab casing
        if let exact = vocab.first(where: { $0.lowercased() == lower }) {
            return word.replacingOccurrences(of: core, with: exact)
        }
        // An ordinary English word is what the speaker meant. Never rewrite it.
        // Without this guard "not" becomes "notes" whenever "notes" is in the vocab list
        // (Jaro-Winkler scores them 0.907), and every such rewrite corrupts a dictation.
        guard !isRealWord(lower) else { return word }
        var best: (term: String, score: Double)?
        for term in vocab {
            let score = similarity(lower, term.lowercased())
            if score > (best?.score ?? 0) { best = (term, score) }
        }
        if let best, best.score >= threshold(forLength: core.count) {
            return word.replacingOccurrences(of: core, with: best.term)
        }
        return word
    }

    /// Two spoken words that together sound like one vocab term ("open whisper" → "OpenWispr").
    /// The halves are deliberately NOT real-word-guarded: both "open" and "whisper" are ordinary
    /// words, and guarding them would defeat the only case this exists for. Instead the *joined*
    /// form must not be a real word ("not able" → "notable" is rejected here), the vocab term must
    /// be long enough to be a proper noun, and the bar is raised to `phraseThreshold`.
    private static func mergedMatch(_ a: String, _ b: String, vocab: [String],
                                    isRealWord: (String) -> Bool) -> String? {
        let trimSet = CharacterSet(charactersIn: ".,!?;:\"'()")
        let coreA = a.trimmingCharacters(in: trimSet)
        let coreB = b.trimmingCharacters(in: trimSet)
        guard coreA.count >= 2, coreB.count >= 2 else { return nil }
        // a merge would swallow punctuation sitting between or before the pair
        guard coreA == a, let rangeB = b.range(of: coreB), rangeB.lowerBound == b.startIndex else { return nil }
        let joined = (coreA + coreB).lowercased()
        guard !isRealWord(joined) else { return nil }
        var best: (term: String, score: Double)?
        for term in vocab where term.count >= 6 {
            let flat = term.replacingOccurrences(of: " ", with: "").lowercased()
            // a merge should be about the same length as its target, not a term plus a stray word
            guard abs(flat.count - joined.count) <= 3 else { continue }
            let score = similarity(joined, flat)
            if score > (best?.score ?? 0) { best = (term, score) }
        }
        // The merge must beat what the first word achieves alone, or "stratsea and" swallows "and":
        // the joined form still scores high because Jaro-Winkler rewards the shared prefix.
        var soloBest = 0.0
        if !isRealWord(coreA.lowercased()) {
            for term in vocab {
                soloBest = max(soloBest, similarity(coreA.lowercased(), term.lowercased()))
            }
        }
        guard let best, best.score >= phraseThreshold, best.score > soloBest else { return nil }
        return best.term + String(b[rangeB.upperBound...])   // keep trailing punctuation
    }

    private static func threshold(forLength n: Int) -> Double {
        n >= 6 ? fuzzyThreshold : shortWordThreshold
    }

    /// Best of the spelled and the sounded match. Whisper mis-hears proper nouns phonetically,
    /// so "stratsea" is only 0.868 against "Stratzy" by spelling but 1.0 by sound.
    static func similarity(_ a: String, _ b: String) -> Double {
        max(jaroWinkler(a, b), jaroWinkler(phoneticKey(a), phoneticKey(b)))
    }

    // MARK: - Phonetic key

    /// Reduces a word to roughly what it sounds like: drop vowels after the first, fold letters
    /// that share a sound (c/q→k, z→s, d→t, b→p, v→f, g→k, x→ks), collapse doubles.
    /// "stratsea" and "Stratzy" both reduce to "strts".
    static func phoneticKey(_ s: String) -> String {
        var w = s.lowercased()
        for (from, to) in [("ph", "f"), ("ck", "k"), ("qu", "kw")] {
            w = w.replacingOccurrences(of: from, with: to)
        }
        let fold: [Character: String] = ["c": "k", "q": "k", "z": "s", "x": "ks",
                                         "v": "f", "g": "k", "d": "t", "b": "p"]
        var folded = ""
        for (i, ch) in w.enumerated() {
            guard ch.isLetter else { continue }
            if "aeiouy".contains(ch) {
                if i == 0 { folded.append(ch) }   // a leading vowel still distinguishes the word
                continue
            }
            folded += fold[ch] ?? String(ch)
        }
        var collapsed = ""
        for ch in folded where collapsed.last != ch { collapsed.append(ch) }
        return collapsed
    }

    // MARK: - Learning from user edits

    /// Is an edit from one word to another a spelling correction worth learning, or a content change?
    /// Shared by `CorrectionsWatcher` (clipboard diff) and `AXEditWatcher` (accessibility diff),
    /// which previously each carried their own copy of a band that got this backwards.
    static func isSpellingFix(_ from: String, _ to: String,
                              isRealWord: (String) -> Bool = defaultIsRealWord) -> Bool {
        let a = from.lowercased(), b = to.lowercased()
        guard from != to else { return false }
        if a == b { return true }                    // case-only fix: "delhi" → "Delhi"
        guard !isRealWord(b) else { return false }   // edited into ordinary English = content change
        return jaroWinkler(a, b) >= 0.70             // floor only; see the guard above
    }

    // MARK: - Real-word guard

    private nonisolated(unsafe) static var realWordCache: [String: Bool] = [:]
    private static let realWordLock = NSLock()

    /// Is this an ordinary English word? Backed by the macOS spell checker, which knows the
    /// inflections `/usr/share/dict/words` omits — that list has no "notes", "nodes" or "models",
    /// which are exactly the words that were being corrupted.
    /// Injected into `apply` so the headless self-test stays deterministic and AppKit-free.
    static func defaultIsRealWord(_ word: String) -> Bool {
        let key = word.lowercased()
        realWordLock.lock()
        let cached = realWordCache[key]
        realWordLock.unlock()
        if let cached { return cached }
        #if canImport(AppKit)
        // checkSpelling returns the range of the first misspelling; NSNotFound means it is a word.
        let known = NSSpellChecker.shared.checkSpelling(of: key, startingAt: 0).location == NSNotFound
        #else
        let known = false
        #endif
        realWordLock.lock()
        realWordCache[key] = known
        realWordLock.unlock()
        return known
    }

    // MARK: - Jaro-Winkler

    static func jaroWinkler(_ a: String, _ b: String) -> Double {
        let j = jaro(a, b)
        // common prefix up to 4
        let ca = Array(a), cb = Array(b)
        var prefix = 0
        for i in 0..<min(4, min(ca.count, cb.count)) {
            if ca[i] == cb[i] { prefix += 1 } else { break }
        }
        return j + Double(prefix) * 0.1 * (1 - j)
    }

    private static func jaro(_ a: String, _ b: String) -> Double {
        let s1 = Array(a), s2 = Array(b)
        if s1.isEmpty && s2.isEmpty { return 1 }
        if s1.isEmpty || s2.isEmpty { return 0 }
        let matchDistance = max(s1.count, s2.count) / 2 - 1
        var s1Matches = [Bool](repeating: false, count: s1.count)
        var s2Matches = [Bool](repeating: false, count: s2.count)
        var matches = 0
        for i in 0..<s1.count {
            let start = max(0, i - matchDistance)
            let end = min(i + matchDistance + 1, s2.count)
            guard start < end else { continue }
            for k in start..<end where !s2Matches[k] && s1[i] == s2[k] {
                s1Matches[i] = true; s2Matches[k] = true; matches += 1; break
            }
        }
        if matches == 0 { return 0 }
        var t = 0.0, k = 0
        for i in 0..<s1.count where s1Matches[i] {
            while !s2Matches[k] { k += 1 }
            if s1[i] != s2[k] { t += 0.5 }
            k += 1
        }
        let m = Double(matches)
        return (m / Double(s1.count) + m / Double(s2.count) + (m - t) / m) / 3
    }

    /// Driven by `core/fixtures/dictionary.json`, the same table the Rust port is held to.
    /// Every threshold here was tuned against a real dictation failure, so the cases are
    /// the contract: a change that moves any of them changes what users get.
    private struct FixtureFile: Decodable {
        struct Pair: Decodable {
            let from: String
            let to: String
        }
        struct ApplyCase: Decodable {
            let vocab: [String]?
            let replacements: [Pair]?
            let input: String
            let expected: String
        }
        struct ScoreCase: Decodable {
            let a: String
            let b: String
            let min: Double?
            let max: Double?
        }
        struct FixCase: Decodable {
            let from: String
            let to: String
            let expected: Bool
        }
        /// Stub English so the gate is deterministic and needs no AppKit spell checker.
        let english: [String]
        let apply: [ApplyCase]
        let jaroWinkler: [ScoreCase]
        let phoneticEqual: [[String]]
        let phoneticDiffer: [[String]]
        let spellingFix: [FixCase]
    }

    static func selfTest() {
        let f = Fixtures.load(FixtureFile.self, "dictionary.json")
        let english = Set(f.english.map { $0.lowercased() })
        let stub: (String) -> Bool = { english.contains($0.lowercased()) }

        Fixtures.expect(!f.apply.isEmpty, "dictionary.json has no apply cases")
        for c in f.apply {
            var d = DictionaryData()
            d.vocab = c.vocab ?? []
            d.replacements = (c.replacements ?? []).map { Replacement(from: $0.from, to: $0.to) }
            Fixtures.expectEqual(apply(c.input, d, isRealWord: stub), c.expected, "apply(\(c.input))")
        }
        for c in f.jaroWinkler {
            let score = jaroWinkler(c.a, c.b)
            if let min = c.min { Fixtures.expect(score > min, "jw(\(c.a), \(c.b)) = \(score), want > \(min)") }
            if let max = c.max { Fixtures.expect(score < max, "jw(\(c.a), \(c.b)) = \(score), want < \(max)") }
        }
        for pair in f.phoneticEqual {
            Fixtures.expectEqual(phoneticKey(pair[0]), phoneticKey(pair[1]), "phonetic keys should match")
        }
        for pair in f.phoneticDiffer {
            Fixtures.expect(phoneticKey(pair[0]) != phoneticKey(pair[1]),
                            "phonetic keys should differ: \(pair[0]) vs \(pair[1])")
        }
        for c in f.spellingFix {
            Fixtures.expectEqual(isSpellingFix(c.from, c.to, isRealWord: stub), c.expected,
                                 "isSpellingFix(\(c.from), \(c.to))")
        }
        print("DictionaryStore.selfTest PASS (\(f.apply.count) apply cases)")
    }
}

File: core/src/dictionary.rs
//! The user dictionary: exact phrase replacements, then fuzzy correction of near-miss
//! words toward the user's custom vocabulary.
//!
//! Ported from `Dictionary.swift`. Every threshold and guard here was tuned against real
//! dictation failures, so `fixtures/dictionary.json` pins the behaviour on both platforms.
//! The real-word oracle is injected rather than owned: macOS uses `NSSpellChecker`,
//! Windows uses `ISpellChecker`, and tests use a fixed word list.

use regex::RegexBuilder;
use serde::{Deserialize, Serialize};

/// Words of 6 or more characters.
pub const FUZZY_THRESHOLD: f64 = 0.86;
/// 3 to 5 characters: one edit is a large share of a short word.
pub const SHORT_WORD_THRESHOLD: f64 = 0.92;
/// Two spoken words merged into one vocabulary term.
pub const PHRASE_THRESHOLD: f64 = 0.92;

const TRIM: &[char] = &['.', ',', '!', '?', ';', ':', '"', '\'', '(', ')'];

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct Replacement {
    pub from: String,
    pub to: String,
}

#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize, Deserialize)]
pub struct DictionaryData {
    /// Preferred spellings, fuzzy-matched (for example "WhisperKit").
    #[serde(default)]
    pub vocab: Vec<String>,
    /// Exact phrase to replacement.
    #[serde(default)]
    pub replacements: Vec<Replacement>,
}

/// Answers "is this an ordinary English word?".
pub type IsRealWord<'a> = &'a dyn Fn(&str) -> bool;

pub fn apply(text: &str, d: &DictionaryData, is_real_word: IsRealWord) -> String {
    let mut s = text.to_string();

    // 1) exact phrase replacements, case-insensitive
    for r in d.replacements.iter().filter(|r| !r.from.is_empty()) {
        let pattern = format!(r"\b{}\b", regex::escape(&r.from));
        if let Ok(re) = RegexBuilder::new(&pattern).case_insensitive(true).build() {
            s = re.replace_all(&s, r.to.as_str()).into_owned();
        }
    }

    // 2) fuzzy-correct words toward the custom vocabulary
    if d.vocab.is_empty() {
        return s;
    }
    let words: Vec<&str> = s.split(' ').collect();
    let mut out: Vec<String> = Vec::with_capacity(words.len());
    let mut i = 0;
    while i < words.len() {
        // Prefer a two-word match, so "open whisper" beats correcting each half separately.
        if i + 1 < words.len() {
            if let Some(merged) = merged_match(words[i], words[i + 1], &d.vocab, is_real_word) {
                out.push(merged);
                i += 2;
                continue;
            }
        }
        out.push(correct_word(words[i], &d.vocab, is_real_word));
        i += 1;
    }
    out.join(" ")
}

fn correct_word(word: &str, vocab: &[String], is_real_word: IsRealWord) -> String {
    // preserve leading and trailing punctuation
    let core = word.trim_matches(|c| TRIM.contains(&c));
    if core.chars().count() < 3 {
        return word.to_string();
    }
    let lower = core.to_lowercase();

    // already correct (case-insensitive exact) → snap to the vocabulary's casing
    if let Some(exact) = vocab.iter().find(|v| v.to_lowercase() == lower) {
        return word.replace(core, exact);
    }

    // An ordinary English word is what the speaker meant. Never rewrite it.
    // Without this guard "not" becomes "notes" whenever "notes" is in the vocabulary
    // (Jaro-Winkler scores them 0.907), and every such rewrite corrupts a dictation.
    if is_real_word(&lower) {
        return word.to_string();
    }

    let mut best: Option<(&String, f64)> = None;
    for term in vocab {
        let score = similarity(&lower, &term.to_lowercase());
        if score > best.map_or(0.0, |(_, s)| s) {
            best = Some((term, score));
        }
    }
    if let Some((term, score)) = best {
        if score >= threshold_for_length(core.chars().count()) {
            return word.replace(core, term);
        }
    }
    word.to_string()
}

/// Two spoken words that together sound like one vocabulary term ("open whisper" → "OpenWispr").
///
/// The halves are deliberately NOT real-word-guarded: both "open" and "whisper" are ordinary
/// words, and guarding them would defeat the only case this exists for. Instead the *joined*
/// form must not be a real word ("not able" → "notable" is rejected here), the vocabulary term
/// must be long enough to be a proper noun, and the bar is raised to `PHRASE_THRESHOLD`.
fn merged_match(a: &str, b: &str, vocab: &[String], is_real_word: IsRealWord) -> Option<String> {
    let core_a = a.trim_matches(|c| TRIM.contains(&c));
    let core_b = b.trim_matches(|c| TRIM.contains(&c));
    if core_a.chars().count() < 2 || core_b.chars().count() < 2 {
        return None;
    }
    // a merge would swallow punctuation sitting between or before the pair
    if core_a != a || !b.starts_with(core_b) {
        return None;
    }
    let joined = format!("{core_a}{core_b}").to_lowercase();
    if is_real_word(&joined) {
        return None;
    }
    let joined_len = joined.chars().count() as i64;

    let mut best: Option<(&String, f64)> = None;
    for term in vocab.iter().filter(|t| t.chars().count() >= 6) {
        let flat = term.replace(' ', "").to_lowercase();
        // a merge should be about the same length as its target, not a term plus a stray word
        if (flat.chars().count() as i64 - joined_len).abs() > 3 {
            continue;
        }
        let score = similarity(&joined, &flat);
        if score > best.map_or(0.0, |(_, s)| s) {
            best = Some((term, score));
        }
    }

    // The merge must beat what the first word achieves alone, or "stratsea and" swallows "and":
    // the joined form still scores high because Jaro-Winkler rewards the shared prefix.
    let mut solo_best = 0.0_f64;
    let lower_a = core_a.to_lowercase();
    if !is_real_word(&lower_a) {
        for term in vocab {
            solo_best = solo_best.max(similarity(&lower_a, &term.to_lowercase()));
        }
    }

    let (term, score) = best?;
    if score < PHRASE_THRESHOLD || score <= solo_best {
        return None;
    }
    // keep trailing punctuation
    Some(format!("{}{}", term, &b[core_b.len()..]))
}

fn threshold_for_length(n: usize) -> f64 {
    if n >= 6 {
        FUZZY_THRESHOLD
    } else {
        SHORT_WORD_THRESHOLD
    }
}

/// Best of the spelled and the sounded match. Whisper mis-hears proper nouns phonetically,
/// so "stratsea" is only 0.868 against "Stratzy" by spelling but 1.0 by sound.
pub fn similarity(a: &str, b: &str) -> f64 {
    jaro_winkler(a, b).max(jaro_winkler(&phonetic_key(a), &phonetic_key(b)))
}

/// Reduces a word to roughly what it sounds like: drop vowels after the first, fold letters
/// that share a sound (c/q→k, z→s, d→t, b→p, v→f, g→k, x→ks), collapse doubles.
/// "stratsea" and "Stratzy" both reduce to "strts".
pub fn phonetic_key(s: &str) -> String {
    let mut w = s.to_lowercase();
    for (from, to) in [("ph", "f"), ("ck", "k"), ("qu", "kw")] {
        w = w.replace(from, to);
    }
    let mut folded = String::with_capacity(w.len());
    for (i, ch) in w.chars().enumerate() {
        if !ch.is_alphabetic() {
            continue;
        }
        if "aeiouy".contains(ch) {
            // a leading vowel still distinguishes the word
            if i == 0 {
                folded.push(ch);
            }
            continue;
        }
        match ch {
            'c' | 'q' | 'g' => folded.push('k'),
            'z' => folded.push('s'),
            'x' => folded.push_str("ks"),
            'v' => folded.push('f'),
            'd' => folded.push('t'),
            'b' => folded.push('p'),
            other => folded.push(other),
        }
    }
    let mut collapsed = String::with_capacity(folded.len());
    for ch in folded.chars() {
        if !collapsed.ends_with(ch) {
            collapsed.push(ch);
        }
    }
    collapsed
}

/// Is an edit from one word to another a spelling correction worth learning, or a content change?
/// Shared by the clipboard diff and the accessibility diff, which each used to carry their own
/// copy of a band that got this backwards.
pub fn is_spelling_fix(from: &str, to: &str, is_real_word: IsRealWord) -> bool {
    if from == to {
        return false;
    }
    let a = from.to_lowercase();
    let b = to.to_lowercase();
    if a == b {
        return true; // case-only fix: "delhi" → "Delhi"
    }
    if is_real_word(&b) {
        return false; // edited into ordinary English = content change
    }
    jaro_winkler(&a, &b) >= 0.70 // floor only; see the guard above
}

pub fn jaro_winkler(a: &str, b: &str) -> f64 {
    let j = jaro(a, b);
    let ca: Vec<char> = a.chars().collect();
    let cb: Vec<char> = b.chars().collect();
    // common prefix up to 4
    let mut prefix = 0;
    for i in 0..ca.len().min(cb.len()).min(4) {
        if ca[i] == cb[i] {
            prefix += 1;
        } else {
            break;
        }
    }
    j + prefix as f64 * 0.1 * (1.0 - j)
}

fn jaro(a: &str, b: &str) -> f64 {
    let s1: Vec<char> = a.chars().collect();
    let s2: Vec<char> = b.chars().collect();
    if s1.is_empty() && s2.is_empty() {
        return 1.0;
    }
    if s1.is_empty() || s2.is_empty() {
        return 0.0;
    }
    // Swift computes this with Int division, so short strings get a negative window and
    // therefore zero matches. Kept as isize so the two implementations agree.
    let match_distance = (s1.len().max(s2.len()) / 2) as isize - 1;
    let mut s1_matches = vec![false; s1.len()];
    let mut s2_matches = vec![false; s2.len()];
    let mut matches = 0usize;

    for i in 0..s1.len() {
        let start = (i as isize - match_distance).max(0);
        let end = (i as isize + match_distance + 1).min(s2.len() as isize);
        if start >= end {
            continue;
        }
        for k in start as usize..end as usize {
            if !s2_matches[k] && s1[i] == s2[k] {
                s1_matches[i] = true;
                s2_matches[k] = true;
                matches += 1;
                break;
            }
        }
    }
    if matches == 0 {
        return 0.0;
    }

    let mut t = 0.0;
    let mut k = 0usize;
    for i in 0..s1.len() {
        if !s1_matches[i] {
            continue;
        }
        while !s2_matches[k] {
            k += 1;
        }
        if s1[i] != s2[k] {
            t += 0.5;
        }
        k += 1;
    }

    let m = matches as f64;
    (m / s1.len() as f64 + m / s2.len() as f64 + (m - t) / m) / 3.0
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::fixtures;
    use std::collections::HashSet;

    #[derive(Deserialize)]
    struct DictFixtures {
        /// Stub English so the gate is deterministic and needs no system spell checker.
        english: Vec<String>,
        apply: Vec<ApplyCase>,
        jaro_winkler: Vec<ScoreCase>,
        phonetic_equal: Vec<[String; 2]>,
        phonetic_differ: Vec<[String; 2]>,
        spelling_fix: Vec<FixCase>,
    }

    #[derive(Deserialize)]
    struct ApplyCase {
        #[serde(default)]
        vocab: Vec<String>,
        #[serde(default)]
        replacements: Vec<Replacement>,
        input: String,
        expected: String,
    }

    #[derive(Deserialize)]
    struct ScoreCase {
        a: String,
        b: String,
        #[serde(default)]
        min: Option<f64>,
        #[serde(default)]
        max: Option<f64>,
    }

    #[derive(Deserialize)]
    struct FixCase {
        from: String,
        to: String,
        expected: bool,
    }

    #[test]
    fn matches_the_shared_fixtures() {
        let f: DictFixtures = fixtures::load("dictionary.json");
        let english: HashSet<String> = f.english.iter().map(|w| w.to_lowercase()).collect();
        let is_real = |w: &str| english.contains(&w.to_lowercase());

        assert!(!f.apply.is_empty(), "fixture file is empty");
        for case in &f.apply {
            let d = DictionaryData {
                vocab: case.vocab.clone(),
                replacements: case.replacements.clone(),
            };
            let got = apply(&case.input, &d, &is_real);
            assert_eq!(got, case.expected, "apply({:?})", case.input);
        }
        for case in &f.jaro_winkler {
            let score = jaro_winkler(&case.a, &case.b);
            if let Some(min) = case.min {
                assert!(
                    score > min,
                    "jw({}, {}) = {score}, want > {min}",
                    case.a,
                    case.b
                );
            }
            if let Some(max) = case.max {
                assert!(
                    score < max,
                    "jw({}, {}) = {score}, want < {max}",
                    case.a,
                    case.b
                );
            }
        }
        for [a, b] in &f.phonetic_equal {
            assert_eq!(
                phonetic_key(a),
                phonetic_key(b),
                "phonetic keys should match"
            );
        }
        for [a, b] in &f.phonetic_differ {
            assert_ne!(
                phonetic_key(a),
                phonetic_key(b),
                "phonetic keys should differ"
            );
        }
        for case in &f.spelling_fix {
            assert_eq!(
                is_spelling_fix(&case.from, &case.to, &is_real),
                case.expected,
                "is_spelling_fix({:?}, {:?})",
                case.from,
                case.to
            );
        }
    }
}
